### PR TITLE
chore: switch nyc to c8

### PR DESCRIPTION
**Pull request by aladdin-add**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ x] Other, please explain:

#### What changes did you make? (Give an overview)

change nyc => c8, just like we had done in other eslint repos like eslint/eslintrc, eslint/espree.

#### Is there anything you'd like reviewers to focus on?
I encountered an issue: https://github.com/bcoe/c8/issues/411, but seems it has been fixed after upgrade c8 to v7.12.0(latest).

### CODE REVIEW

**Review by mdjermanovic — COMMENTED** [review]
The coverage summary looks significantly different?

This branch (c8):

```
=============================== Coverage summary ===============================
Statements   : 99.09% ( 76580/77279 )
Branches     : 98.27% ( 14688/14946 )
Functions    : 99% ( 3200/3232 )
Lines        : 99.09% ( 76580/77279 )
================================================================================
```

main branch (nyc):

```
=============================== Coverage summary ===============================
Statements   : 98.53% ( 16639/16886 )
Branches     : 97.31% ( 13882/14265 )
Functions    : 99.27% ( 3717/3744 )
Lines        : 98.53% ( 16367/16611 )
================================================================================
```

**Comment by aladdin-add** [comment]
not sure why, I'll try to investigate it later.

**Comment by aladdin-add** [comment]
nyc:
image

c8:
image

I had compared the 2 result: seems c8 has differently counted linebreaks/comments(as it's using [Node.js' built in functionality](https://nodejs.org/dist/latest-v10.x/docs/api/cli.html#cli_node_v8_coverage_dir)). another issue is related to v8 itself: https://github.com/bcoe/c8/issues/227

**Comment by nzakas** [comment]
Yeah, nyc uses instrumentation, which can affect the results. I think c8 is likely more accurate than nyc was.

**Review by nzakas — APPROVED** [review]
LGTM. Just want another set of eyes to review before merging.

File: lib/linter/code-path-analysis/code-path-segment.js
Comment (AriPerkkio) [inline lib/linter/code-path-analysis/code-path-segment.js:94]: `c8` shouldn't understand whether there is an `if` statement or not, as it does not parse the AST. There is no such option as `c8 ignore if`.

```suggestion
        /* c8 ignore next 3 */
        if (debug.enabled) {
            this.internal.nodes = [];
        }
```

**Review by AriPerkkio — COMMENTED** [review]
> Yeah, nyc uses instrumentation, which can affect the results. I think c8 is likely more accurate than nyc was.

Just out of curiosity - more accurate in what way? With istanbul it's possible to be very precise where exactly the coverage is wanted to collect from. Native v8 coverage doesn't provide such control. Also processing the v8's results seems very difficult and is thus limited.

There is a great summary from jest regarding istanbul -> c8 migrations, https://github.com/facebook/jest/issues/11188. This explains some of the coverage differences between the two.

File: lib/linter/code-path-analysis/code-path-segment.js
Comment (aladdin-add) [reply lib/linter/code-path-analysis/code-path-segment.js]: thanks, I updated it to use `c8 ignore start/end` - in case it may have > 3 statements in future changes. :)

**Comment by aladdin-add** [comment]
@AriPerkkio One thing I should have mentioned earilier: the main reason migrating to c8 is we want to migrate eslint codebase to esm: https://github.com/eslint/eslint/issues/15560

File: lib/linter/code-path-analysis/code-path-segment.js
Comment (AriPerkkio) [reply lib/linter/code-path-analysis/code-path-segment.js]: This same comment applies to `c8 ignore else` comments. There is one still remaining. 😄 

See https://github.com/bcoe/c8/#ignoring-uncovered-lines-functions-and-blocks for all options.

File: lib/linter/code-path-analysis/code-path-state.js
Comment (AriPerkkio) [inline lib/linter/code-path-analysis/code-path-state.js:62]: I'm not sure whether additional text inside comments is supported at all. Looks like the regular expressions used are quite strict. https://github.com/istanbuljs/v8-to-istanbul/blob/v9.0.1/lib/source.js#L53

Similar fixes are likely required in other parts of PR:

```suggestion
    /* c8 ignore next */
```

**Review by nzakas — APPROVED** [review]

File: lib/linter/code-path-analysis/code-path-state.js
Comment (aladdin-add) [reply lib/linter/code-path-analysis/code-path-state.js]: good catch! should be addressed in https://github.com/eslint/eslint/pull/16263/commits/82b3ded068dfa9b1b695f482f46db875b84de9ed.

**Review by AriPerkkio — APPROVED** [review]

**Pull request merged**
